src/librbd/operation/DisableFeaturesRequest.cc

```diff
@@ -636,16 +649,6 @@ Context *DisableFeaturesRequest<I>::handle_finish(int r) {
   CephContext *cct = image_ctx.cct;
   ldout(cct, 20) << this << " " << __func__ << ": r=" << r << dendl;
 
-  {
-    std::unique_lock locker{image_ctx.owner_lock};
-    if (image_ctx.exclusive_lock != nullptr && m_requests_blocked) {
-      image_ctx.exclusive_lock->unblock_requests();
-    }
-
-    image_ctx.io_image_dispatcher->unblock_writes();
-  }
-  image_ctx.state->handle_prepare_lock_complete();
```

Comment:
With this change, now if the request fails, e.g. in "remove_object_map", this code will not execute, i.e. it will not unblock lock requests and writes.

I suppose you should see this when running the mock tests, which need update too, after this change BTW.

And you need to update the state diagram in  Enable/DisableFeaturesRequest.h  files accordingly.

Reply:
> I suppose you should see this when running the mock tests, which need update too, after this change BTW.

```
[pkalever@localhost build]$ ./bin/unittest_librbd --gtest_filter=TestMockOperationEnableFeaturesRequest*
seed 2125062
Note: Google Test filter = TestMockOperationEnableFeaturesRequest*
[==========] Running 7 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 7 tests from TestMockOperationEnableFeaturesRequest
[ RUN      ] TestMockOperationEnableFeaturesRequest.All
SKIPPING
[       OK ] TestMockOperationEnableFeaturesRequest.All (1 ms)
[ RUN      ] TestMockOperationEnableFeaturesRequest.ObjectMap
SKIPPING
[       OK ] TestMockOperationEnableFeaturesRequest.ObjectMap (0 ms)
[ RUN      ] TestMockOperationEnableFeaturesRequest.ObjectMapError
SKIPPING
[       OK ] TestMockOperationEnableFeaturesRequest.ObjectMapError (0 ms)
[ RUN      ] TestMockOperationEnableFeaturesRequest.SetFlagsError
SKIPPING
[       OK ] TestMockOperationEnableFeaturesRequest.SetFlagsError (1 ms)
[ RUN      ] TestMockOperationEnableFeaturesRequest.Mirroring
SKIPPING
[       OK ] TestMockOperationEnableFeaturesRequest.Mirroring (0 ms)
[ RUN      ] TestMockOperationEnableFeaturesRequest.JournalingError
SKIPPING
[       OK ] TestMockOperationEnableFeaturesRequest.JournalingError (0 ms)
[ RUN      ] TestMockOperationEnableFeaturesRequest.MirroringError
SKIPPING
[       OK ] TestMockOperationEnableFeaturesRequest.MirroringError (0 ms)
[----------] 7 tests from TestMockOperationEnableFeaturesRequest (2 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test suite ran. (28 ms total)
[  PASSED  ] 7 tests.
[pkalever@localhost build]$ ./bin/unittest_librbd --gtest_filter=TestMockOperationDisableFeaturesRequest*
seed 2125171
Note: Google Test filter = TestMockOperationDisableFeaturesRequest*
[==========] Running 5 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 5 tests from TestMockOperationDisableFeaturesRequest
[ RUN      ] TestMockOperationDisableFeaturesRequest.All
SKIPPING
[       OK ] TestMockOperationDisableFeaturesRequest.All (0 ms)
[ RUN      ] TestMockOperationDisableFeaturesRequest.ObjectMap
SKIPPING
[       OK ] TestMockOperationDisableFeaturesRequest.ObjectMap (1 ms)
[ RUN      ] TestMockOperationDisableFeaturesRequest.ObjectMapError
SKIPPING
[       OK ] TestMockOperationDisableFeaturesRequest.ObjectMapError (0 ms)
[ RUN      ] TestMockOperationDisableFeaturesRequest.Mirroring
SKIPPING
[       OK ] TestMockOperationDisableFeaturesRequest.Mirroring (0 ms)
[ RUN      ] TestMockOperationDisableFeaturesRequest.MirroringError
SKIPPING
[       OK ] TestMockOperationDisableFeaturesRequest.MirroringError (0 ms)
[----------] 5 tests from TestMockOperationDisableFeaturesRequest (1 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test suite ran. (28 ms total)
[  PASSED  ] 5 tests.
[pkalever@localhost build]$ 
```

I hope you mean `TestMockOperationEnableFeaturesRequest` and `TestMockOperationDisableFeaturesRequest`. I just ran both and they work as expected with the latest changes.

I have incorporated the rest of the above suggestions, PTAL.

Reply:
Many of the tests were skipped. Try `RBD_FEATURES=127 ./bin/unittest_librbd ...`

Reply:
Done now. thanks!

Reply:
> Many of the tests were skipped.

I guess we should make that more obvious...  The confusing part is that in the end these tests are reported as `PASSED`.  I'll make it so that they are reported is `SKIPPED` and are counted appropriately.

Reply:
I agree!